master/pkg/schemas/expconf/hyperparameters_config.go

```diff
@@ -12,7 +13,117 @@ const GlobalBatchSize = "global_batch_size"
 
 //go:generate ../gen.sh
 // HyperparametersV0 is a versioned hyperparameters config.
-type HyperparametersV0 map[string]HyperparameterV0
+type HyperparametersV0 map[string]interface{}
+
+// FlattenHPs returns a flat dictionary with keys representing nested structure.
+// For example, {"optimizer": {"learning_rate": 0.01}} will be flattened to
+// {"optimizer.learning_rate": 0.01}.
+func FlattenHPs(h HyperparametersV0) HyperparametersV0 {
+	flatHPs := make(HyperparametersV0)
+	flattenHPs(h, "", flatHPs)
+	return flatHPs
+}
+
+func flattenHPs(h HyperparametersV0, prefix string, target HyperparametersV0) {
+	for key, element := range h {
+		switch tHP := element.(type) {
+		case HyperparametersV0:
+			flattenHPs(tHP, prefix+key+".", target)
+		default:
+			target[prefix+key] = element
+		}
+	}
+}
+
+// UnflattenHPs undos the FlattenHPs function to restore the nested structure.
+func UnflattenHPs(h HyperparametersV0) HyperparametersV0 {
```

Comment:
This seems awfully dangerous, because after mangling a user namespace (hyperparameter key names) we are now attempting to unmangle it and the system will try to operate on the result.

Mangling it for e.g. display purposes (in the webui maybe) is reasonable, but the mangled version should only ever be used for cosmetic things.

Reply:
FWIW, we did run into this exactly at some point in the master config, viper mangles configs exactly the same as is done here and so `.` in a podspec was unacceptable (even though it's in keys frequently). You could just use a more unique pattern to mangle (I changed viper to use `..`), but prob not mangling at all is better.

Reply:
Thanks.  I removed `UnflattenHPs` but still use an internal method `unflattenSample` for generating the hp grid in grid search.  I did adopt `..` as the separator for more uniqueness.